### PR TITLE
cache-url-builder

### PR DESCRIPTION
**Pull request by chenshay**

### CODE REVIEW

File: src/cache-url-builder.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+/**
+ * Converts a Unicode string representing a domain name to punycode. Only the
+ * non-ASCII parts of the domain name will be converted, i.e. it doesn't
+ * matter if you call it with a domain that's already in ASCII.
+ *
+ * @param {string} domain The domain name to convert, as a Unicode string.
+ * @return {string} The punycode representation of the given domain name.
+ */
+function toAscii(domain) {
```
Comment (ericfs) [inline src/cache-url-builder.js]: private?

File: src/cache-url-builder.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+ * 
+ * Fallback Algorithm:
+ * 1. Take the SHA256 of the punycode view of the domain.
+ * 2. Base32 encode the resulting hash. Set the domain prefix to the resulting
+ *    string.
+ *
+ * @param {string} domain The publisher domain
+ * @return {string} The curls encoded domain
+ */
+export function constructCacheUrl(domain) {
```
Comment (ericfs) [inline src/cache-url-builder.js]: This name is a bit confusing. It is just constructing the cache domain not the complete URL. You'll need the complete URL, but having a function that just constructs the domain seems reasonable.

File: src/cache-url-builder.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+ * 1. Convert pub.com from IDN (punycode) to utf-8, if applicable.
+ * 2. Replace every “-” with “--”.
+ * 3. Replace each “.” with “-”.
+ * 4. Convert back to IDN.
+ * 
+ * Examples:
+ *   'something.com'    =>  'something-com'
+ *   'SOMETHING.COM'    =>  'something-com'
+ *   'hello-world.com'  =>  'hello--world-com'
+ *   'hello--world.com' =>  'hello----world-com'
```
Comment (ericfs) [inline src/cache-url-builder.js]: It would be good to have these enumerated with actual test cases. There are a variety of edge cases that should be tested.

File: src/cache-url-builder.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+/**
+ * A generic Array map utility function.
+ *
+ * @param {Array} array The array to iterate over.
+ * @param {Function} callback The function that gets called for every array
+ *   item.
+ * @return {Array} A new array of values returned by the callback function.
+ * @private
+ */
+function map_(array, callback) {
```
Comment (ericfs) [inline src/cache-url-builder.js]: why not just Array.map()?

File: src/cache-url-builder.js (hunk 3 above)
Comment (chenshay) [inline src/cache-url-builder.js]: yes - we still need to add test capabilities to this project.
@erwinmombay

File: src/cache-url-builder.js (hunk 2 above)
Comment (chenshay) [inline src/cache-url-builder.js]: I haven't yet decided if I want the complete url to be a separate method that calls this one or if I want to do it all in 1 go. Thoughts?

Comment (ericfs) [inline src/cache-url-builder.js]: I would think you'll want one function to return the complete URL. I don't have a strong opinion if that function should call this function or whether you should just have one function.

Comment (chenshay) [inline src/cache-url-builder.js]: SG. Added TODO to return complete url

File: src/cache-url-builder.js [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ * @param {string} domain The publisher domain
+ * @return {string} The curls encoded domain
+ * @private
+ */
+function constructHumanReadableCurlsProxyDomain_(domain) {
+  domain = punycode.toUnicode(domain);
+  domain = domain.split('-').join('--');
+  domain = domain.split('.').join('-');
+  return punycode.toASCII(domain).toLowerCase();;
```
Comment (ericfs) [inline src/cache-url-builder.js]: nit: extra ;

File: src/cache-url-builder.js [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+  '\u0591-\u06EF\u06FA-\u07FF\u200F\uFB1D-\uFDFF\uFE70-\uFEFC';
+
+/** @type {RegExp} */
+const HAS_LTR_CHARS = new RegExp('[' + LTR_CHARS + ']');
+
+/** @type {RegExp} */
+const HAS_RTL_CHARS = new RegExp('[' + RTL_CHARS + ']');
+
+/** @type {RegExp} */
+const IS_PUNYCODE = /^xn--/;
```
Comment (ericfs) [inline src/cache-url-builder.js]: not used

File: src/cache-url-builder.js [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+/** @type {RegExp} */
+const SEPARATORS = /\x2E|\u3002|\uFF0E|\uFF61/g;
```
Comment (ericfs) [inline src/cache-url-builder.js]: not used

File: src/cache-url-builder.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+/** @type {RegExp} */
+const NON_ASCII = /[^ -~]/;
```
Comment (ericfs) [inline src/cache-url-builder.js]: not used

File: src/cache-url-builder.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+ * limitations under the License.
+ */
+
+
+const punycode = require('punycode');
+
+/** @type {string} */
+const LTR_CHARS =
+  'A-Za-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02B8\u0300-\u0590\u0800-\u1FFF' +
+  '\u200E\u2C00-\uFB1C\uFE00-\uFE6F\uFEFD-\uFFFF';
```
Comment (ericfs) [inline src/cache-url-builder.js]: Not sure if it exists anywhere, but it would be nice if you could require this instead of hardcoding it here.

Comment (chenshay) [inline src/cache-url-builder.js]: I couldn't find a thing to include unfortunately...

File: src/cache-url-builder.js (hunk 5 above)
Comment (chenshay) [inline src/cache-url-builder.js]: done

File: src/cache-url-builder.js (hunk 8 above)
Comment (chenshay) [inline src/cache-url-builder.js]: done

File: src/cache-url-builder.js (hunk 7 above)
Comment (chenshay) [inline src/cache-url-builder.js]: done

File: src/cache-url-builder.js (hunk 6 above)
Comment (chenshay) [inline src/cache-url-builder.js]: done
